[Coupons] Search for categories

hichamboushaba opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #6362
<!-- Id number of the GitHub issue this PR addresses. -->

Please make sure https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2437 is merged before merging this one.

### Description
This PR adds support for searching for product categories, and it adds unit tests for the main components of the screen.

### Testing instructions
1. Make sure coupons are enabled.
2. Open a coupon in the app.
3. Click on the menu button.
4. Click on edit coupon.
5. Click on "Add categories" or "Select categories".
6. Search for categories in the search view.
7. Confirm the results are retrieved from the API.
8. Confirm child categories are shown too when they are part of the results.

### Images/gif
<img width=400 src="https://user-images.githubusercontent.com/1657201/172811985-421f175f-eb3a-44bd-bb1b-acaaa76f3511.png"/>

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

hichamboushaba committed:
Add a new SearchView component

hichamboushaba committed:
Add implementation for search categories

hichamboushaba committed:
Handle search in the UI layer

hichamboushaba committed:
Improve skeleton view

hichamboushaba committed:
Handle empty screen for search results

hichamboushaba committed:
Add tests for ProductCategoryListHandler

hichamboushaba committed:
Add tests for product categories

hichamboushaba committed:
Show Snackbars on error cases

hichamboushaba committed:
Add more tests

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/compose/component/TextFields.kt:
```diff
@@ -160,6 +172,54 @@ private fun WCOutlinedTextFieldLayout(
 }
 
+@Composable
+fun WCSearchView(
```
@0nko I added this component, I believe you need it to for the product selector.

0nko replied:
Coo, thanks! 👍 I started working on something similar but didn't finish it before the AFK.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/compose/component/TextFields.kt:
```diff
@@ -160,6 +172,54 @@ private fun WCOutlinedTextFieldLayout(
+    modifier: Modifier = Modifier,
+    hint: String = ""
+) {
+    BasicTextField(
```
I opted for using a `BasicTextField` instead of the material `TextField` to be able to fully customize the component, the main difference was the minimum height, `TextField` enforces some padding that makes reducing the `minHeight` not possible.

hichamboushaba committed:
Fix checkstyle issue

hichamboushaba committed:
Bump fluxc hash

hichamboushaba committed:
Fix unit tests

hichamboushaba committed:
Bump fluxc hash

0nko reviewed (commented):
The search looks good! I just have one suggestion - what do you think about adding an 'X' icon button if the text is not empty, which would clear the field?

![image](https://user-images.githubusercontent.com/1522856/173028933-a3a01dc4-fa33-41c2-a24f-922c974ce69d.png)

Also, on iOS the `Clear selection` button is only displayed if there are some items selected. I also did this in product selector.

hichamboushaba committed:
Add a clear button to the search viewtes

hichamboushaba committed:
Hide the clear selection button when no item is selected

hichamboushaba commented:
Thanks for the review @0nko.

> what do you think about adding an 'X' icon button if the text is not empty, which would clear the field?

Good call, I missed this part, I added it in [0e22fd3](https://github.com/woocommerce/woocommerce-android/pull/6722/commits/0e22fd36c3decf8b12eaf1a9d5b9c8db10c64d30)

> Also, on iOS the Clear selection button is only displayed if there are some items selected. I also did this in product selector.

I didn't make it this way because it makes the list jump when "selecting first"/"unselecting last" item:

https://user-images.githubusercontent.com/1657201/173042510-05fb9ab1-948a-4b5a-86bc-a55d505d8735.mov

The cause is that unlike Products where we have the filter button to fill the space always, we don't have anything in Product categories.

To keep things consistent with iOS, I added it now in [f58512e](https://github.com/woocommerce/woocommerce-android/pull/6722/commits/f58512e750d6662cf5489135e558f6a970e308b5), and we can later decide whether it's a good UX or not and apply for all platforms.

0nko reviewed (approved):
Thanks for the updates. Great job! :shipit:
